### PR TITLE
[AIR] Allow user to pass model to `TensorflowCheckpoint.get_model`

### LINKED ISSUE #31290: [AIR] Pass model to `TensorflowCheckpoint.get_model`
### Description

Update `TensorflowCheckpoint.get_model` to support `tf.keras.Model` objects:

**Before**
```
    def get_model(
        self,
        model_definition: Callable[[], tf.keras.Model] | None = None,
    ) -> tf.keras.Model:
```

**After**
```
    def get_model(
        self,
        model: tf.keras.Model | Callable[[], tf.keras.Model]] | None = None,
    ) -> tf.keras.Model:
```

### Use case

Better UX.

When you resume training a TensorFlow model, you may need to create an unnecessary lambda to load model weights. This is because `TensorflowCheckplint.get_model` expects a model definition, but you may have already constructed your model. 

```
def train_loop_per_worker():
    ...

    model = tf.keras.applications.MobileNetV3Small(input_shape=(3, 260, 260))
    if checkpoint is not None:
        model = checkpoint.get_model(lambda: model)
```

### PR DESCRIPTION
**Pull request by bveeramani**
Signed-off-by: Balaji Veeramani <balaji@anyscale.com>

## Why are these changes needed?

When you resume training a TensorFlow model, you may need to create an unnecessary `lambda` to load model weights. This is because `TensorflowCheckplint.get_model` expects a model definition, but you may have already constructed your model. This PR improves the UX by letting users directly pass a model to `get_model`.

**Before**
```
def train_loop_per_worker():
    ...

    model = tf.keras.applications.MobileNetV3Small(input_shape=(3, 260, 260))
    if checkpoint is not None:
        model = checkpoint.get_model(lambda: model)
```

**After**
```
def train_loop_per_worker():
    ...

    model = tf.keras.applications.MobileNetV3Small(input_shape=(3, 260, 260))
    if checkpoint is not None:
        model = checkpoint.get_model(model)
```

## Related issue number

Closes https://github.com/ray-project/ray/issues/31290

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/train/tensorflow/tensorflow_checkpoint.py [hunk 1]
```diff
@@ -210,56 +210,87 @@ def from_saved_model(
...
+                "If you are using BatchPredictor, you should do "
+                "`BatchPredictor.from_checkpoint(checkpoint, TensorflowPredictor)`"
+                " by removing kwargs `model_definition=`. "
+                "If you are using TensorflowPredictor directly, you should do "
+                "`TensorflowPredictor.from_checkpoint(checkpoint)` by "
+                "removing kwargs `model_definition=`."
+            )
+        if self._flavor is self.Flavor.MODEL_WEIGHTS and model is None:
+            raise ValueError(
+                "Expecting `model_definition` argument when checkpoint is "
```
Comment (amogkam) [inline python/ray/train/tensorflow/tensorflow_checkpoint.py:248]: `model_definition` is deprecated now.

File: python/ray/train/tensorflow/tensorflow_checkpoint.py [hunk 2]
```diff
@@ -210,56 +210,87 @@ def from_saved_model(
...
+            model = model_definition
+        if callable(model) and self._flavor is not self.Flavor.MODEL_WEIGHTS:
+            warnings.warn(
+                "TensorflowCheckpoint was created from "
+                "TensorflowCheckpoint.from_saved_model` or "
+                "`TensorflowCheckpoint.from_h5`, which already contains all the "
+                "information needed. This means: "
+                "If you are using BatchPredictor, you should do "
+                "`BatchPredictor.from_checkpoint(checkpoint, TensorflowPredictor)`"
+                " by removing kwargs `model_definition=`. "
```
Comment (amogkam) [inline python/ray/train/tensorflow/tensorflow_checkpoint.py:241]: `model_definition` is now deprecated.

File: python/ray/train/tensorflow/tensorflow_checkpoint.py [hunk 3]
```diff
@@ -210,56 +210,87 @@ def from_saved_model(
...
+                "If you are using TensorflowPredictor directly, you should do "
+                "`TensorflowPredictor.from_checkpoint(checkpoint)` by "
+                "removing kwargs `model_definition=`."
```
Comment (amogkam) [inline python/ray/train/tensorflow/tensorflow_checkpoint.py:244]: `model_definition` is now deprecated.

**Review by amogkam — COMMENTED** [review]
Nice! left some small comments.

File: python/ray/train/tensorflow/tensorflow_checkpoint.py [hunk 4]
```diff
@@ -210,56 +210,87 @@ def from_saved_model(
...
             The Tensorflow Keras model stored in the checkpoint.
         """
+        if model_definition is not None:
+            warnings.warn(
+                "The `model_definition` parameter is deprecated. Use the `model` "
+                "parameter instead.",
+                DeprecationWarning,
+            )
+            model = model_definition
+        if callable(model) and self._flavor is not self.Flavor.MODEL_WEIGHTS:
```
[the remark this answers is not in the document]
Comment (bveeramani) [reply python/ray/train/tensorflow/tensorflow_checkpoint.py:233]: Good catch. Fixed.

**Review by amogkam — APPROVED** [review]

**Pull request merged**